FileInstaller/Installer.cpp

```diff
@@ -0,0 +1,48 @@
+#include <windows.h>
+#include <iostream>
+#include <vector>
+
+#include "FileInstaller.hpp"
+
+Installer::Installer(std::vector<std::shared_ptr<ResourceInstaller>> &resource_installers)
+	:m_resource_installers(resource_installers),
+	m_installed_resource_installers({}),
+	m_is_installed(false) {
+
+	// Validate input doen't have any null values. So there will be no surprises during installation.
+	for (auto& resource_installer : this->m_resource_installers) {
+		if (nullptr == resource_installer) {
+			throw FileInstallerException(FileInstallerStatus::INSTALLER_INVALID_ARGUMENTS);
+		}
+	}
+}
+
+Installer::~Installer() {
+	if (!this->m_is_installed) {
+		this->_revert();
+	}
+}
+
+void Installer::install() {
+	for (auto& resource_installer : this->m_resource_installers) {
+		resource_installer->install();
```

Comment:
There's still an issue here, the push_back below may throw, and no one will remember this action needs to be reverted. 
Think how you can solve it using RAII, without changing the ResourceInstaller interface, or the DIrInstaller and FileInstaller classes.

Reply:
Fixed. Added SafeResourceInstaller